docs/architecture/adr-027-deterministic-protobuf-serialization.md

```diff
@@ -45,11 +46,22 @@ with the following additions:
    must be omitted
 4. `repeated` fields of scalar numeric types must use
    [packed encoding](https://developers.google.com/protocol-buffers/docs/encoding#packed)
-   by default.
-5. Variant encoding of integers must not be longer than needed.
+5. Varint encoding must not be longer than needed:
+    * The highest 6 bits for 10-byte varints must be `0`. 10-byte varints are
+      10 groups of 7 bits, i.e. 70 bits, of which only the lowest 70-6=64 are
+      useful.
+    * No trailing zero bytes (in little endian, i.e. no leading zeroes in big
+      endian). Per rule 3 above, the default value of `0` must be omitted, so
+      this rule does not apply in such cases.
+    * Highest 4 bytes of 8 must be `0` for 32-bit ints. Varints are always
+      64-bit uints, regardless of the numeric type being serialized, but with
```

Comment:
unless I mistunderstood the protobuf spec, this is incorrect.

"Varints are a method of serializing integers using one or more bytes. Smaller numbers take a smaller number of bytes." [source](https://developers.google.com/protocol-buffers/docs/encoding)

Reply:
A sane encoder will use less bytes, sure. But the point of rule 5 (and the point of this PR, to clarify rule 5), is making sane encoding the only valid encoding.

Here's a test that shows protobufjs accepts `2**64-1` as `2^32-1` without a hiccup: https://github.com/lazyledger/protobuf3-solidity-lib/blob/50b97b00bd171a056d7e826c2582da155cfacd26/test/TestFixture.js#L27-L34

Remember, when decoding a varint, it keeps going up to 10 bytes, always. The decode has no idea what the underlying type is, or if it's restricted to 32 bits.

Reply:
Mind you, your lack of understanding of the protobuf spec isn't your fault. Protobuf is simply not specified well, hence the need for this ADR and PR in the first place. That's all the more reason to not use protobuf for canonical serialization, but 🤷